api/envoy/config/route/v3/route_components.proto

```diff
@@ -1564,3 +1567,67 @@ message QueryParameterMatcher {
     bool present_match = 6;
   }
 }
+
+// HTTP Internal Redirect :ref:`architecture overview <arch_overview_internal_redirects>`.
+// [#next-free-field: 6]
+message InternalRedirectPolicy {
+  // Configures :ref:`internal redirect <arch_overview_internal_redirects>` behavior.
+  enum InternalRedirectAction {
+    PASS_THROUGH_INTERNAL_REDIRECT = 0;
+    HANDLE_INTERNAL_REDIRECT = 1;
+  }
+
+  // HTTP scheme combinations that are allowed to be handled as internal
+  // redirect. Downstream scheme comes first and redirect target URL scheme
+  // follows.
```

Comment:
This seems pretty strange to be including the target URL scheme; why isn't is sufficient to include just the downstream? 

For that matter, why do we even need this, given that the route match that leads to the redirect itself has a scheme matcher?

Reply:
This was from the preexisting behavior: it allows a HTTP downstream to be redirected to only HTTP targets; it allows HTTPS downstream to be redirected to any targets.

My understanding is that, if an origin redirect an HTTP request to an HTTPS target, there is a chance that the origin would like to move the downstream to a secure transport for sensitive payload.

While this is not always the case, certainly not our use case, I decided to make this configurable but preserving the existing behavior.

By "scheme matcher" did you mean VirtualHost. require_tls?

Reply:
Scheme seems an odd one here IMO.
upstream scheme is generally just based on "is the connection to upstream secure or not"
I think in general we want to make sure that x-forwarded-proto matches the scheme of the redirect URI.  If we want to say we only redirect https:// requests or http:// requests that'd be the request x-forwarded-proto header not scheme

Reply:
Ah, thanks alyssawilk@ and htuch@!

I think I had misunderstood the purpose of the existing check on x-forwarded-proto vs redirect URI scheme. Our use case led me to think about the redirect URL more as a way to define the desired upstream for a request, but the redirect URL should really be treated purely as a property of the downstream, which makes the existing check make perfect sense, and indeed makes the scheme option here very odd. In other words, Upstream's scheme is completely orthogonal to what the downstream's scheme is.

I'll remove this option completely.

Reply:
This is now done. I removed this option.

I added an option to allow cross scheme redirect - I think this should be fine since as @hutch said, there are scheme match in routes, we can selectively allow certain route to be redirected to, across scheme boundary if we know this is safe.